### PR TITLE
Reorganize integrator functionality in `fem/`

### PR DESCRIPTION
**Pull request by sebastiangrimberg**
The large number of files related to PA and MF bilinear and linear form integrators in `fem/` is getting a bit unwieldy so this PR proposes moving them to a subdirectory `fem/integ/`. It also separates some of the functionality between files with integrator member function definitions and `.hpp` files for the backend kernels, something that had been done a bit for some of the newer functionality but not for others. I also tried to rename files to be a bit more consistent as well across the various integrator and assembly types to help make it more clear for developers.

No new functionality is introduced or attempt to reduce duplicated code is made here, this is just a reorganization of existing methods.

TODO:
- [x] Update the tree structure diagram in [`CONTRIBUTING.md`](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md)
- [x] Document updates in [`CHANGELOG`](https://github.com/mfem/mfem/blob/master/CHANGELOG)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3526](https://github.com/mfem/mfem/pull/3526) | @sebastiangrimberg | @pazner | @pazner + @mlstowell + @YohannDudouit | 4/10/23 | 6/23/23 | ⌛due 5/1/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by mlstowell** [comment]
I think this is a good idea. If we can get some positive feedback from the PA, MF, and EA developers (@YohannDudouit, @camierjs, @dylan-copeland) we should mark this as ready-for-review.

**Comment by YohannDudouit** [comment]
I think this is a very big improvement overall.

**Comment by sebastiangrimberg** [comment]
Thanks @YohannDudouit!

Something that I could also add to this PR would be to complete the move of *all* kernel code out of `.cpp` files where integrator member functions are defined and into `.hpp` files. Right now I only did so in a few places where there are kernels are shared between multiple `.cpp` files, but in general it could be more clear to separate these two types of code everywhere in `fem/integ`. This would come at the expense of making this PR include more changes/new files. Any thoughts here?

**Comment by sebastiangrimberg** [comment]
Seems like this is generally supported so I’ve marked as ready for review. Thank you!

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/integ/bilininteg_diffusion_kernels.hpp
Comment (pazner) [inline fem/integ/bilininteg_diffusion_kernels.hpp:1]: Are the diffs in the file just from rearranging things? It's a bit hard to tell what's changed

File: fem/restriction.cpp [hunk 1]
```diff
@@ -168,8 +168,8 @@ void ElementRestriction::TAddMultTranspose(const Vector& x, Vector& y) const
          for (int j = offset; j < next_offset; ++j)
          {
             const int idx_j = (d_indices[j] >= 0) ? d_indices[j] : -1 - d_indices[j];
-            dof_value += ((d_indices[j] >= 0) ? d_x(idx_j % nd, c, idx_j / nd) :
-                          -d_x(idx_j % nd, c, idx_j / nd));
+            dof_value += (d_indices[j] >= 0) ? d_x(idx_j % nd, c, idx_j / nd) :
+            -d_x(idx_j % nd, c, idx_j / nd);
```
Comment (pazner) [inline fem/restriction.cpp:172]: The old version seems more clearly indented to me

File: linalg/vector.hpp [hunk 2]
```diff
@@ -323,6 +323,9 @@ class Vector
    /// (*this) = -(*this)
    void Neg();
 
+   /// (*this)(i) = 1.0 / (*this)(i)
+   void Reciprocal();
```
Comment (pazner) [inline linalg/vector.hpp:327]: 👍

**Review by pazner — COMMENTED** [review]
Definitely better organization! I didn't check the diffs line-by-line of course, but the structure looks good.

Re: moving kernels from cpp to hpp files, I think we should only do that if there is an immediate need to reuse some of the kernels elsewhere in the library (e.g. the apply mass kernels are also reused in the DG mass CG kernel). So I don't see a particular need to do that in this PR.

File: fem/integ/bilininteg_diffusion_kernels.hpp
Comment (sebastiangrimberg) [reply fem/integ/bilininteg_diffusion_kernels.hpp:1]: Yes, I believe this is just `git diff` not doing a great job at tracking the changes across files. In this case, it tagged `fem/integ/`.hpp` as a rename of `fem/bilininteg_diffusion_pa.cpp`, whereas really the original file was moved mostly to `fem/integ/bilininteg_diffusion_pa.cpp` with some methods going into `bilininteg_diffusion_kernels.hpp`.

I realize this makes a line by line review quite annoying so let me know if I can help in any way. The commits break things down a little bit but in general it is just a reorganization and I haven't done anything beyond copy-pasting functions into their new files.

File: fem/restriction.cpp (hunk 1 above)
Comment (sebastiangrimberg) [reply fem/restriction.cpp]: I agree, this seems to have been a weird `astyle` behavior which got triggered by removing the redundant outer parentheses. Reverted in [f75ece6](https://github.com/mfem/mfem/pull/3526/commits/f75ece6427d65532bd6b954f0c8bb6a697f3fbd2).

**Comment by sebastiangrimberg** [comment]
> Definitely better organization! I didn't check the diffs line-by-line of course, but the structure looks good.
> 
> Re: moving kernels from cpp to hpp files, I think we should only do that if there is an immediate need to reuse some of the kernels elsewhere in the library (e.g. the apply mass kernels are also reused in the DG mass CG kernel). So I don't see a particular need to do that in this PR.

Thank you! Agreed about any further moving of kernels into `hpp` files. For now this PR gets a lot of the reuse and so should be fine as is.

**Review by pazner — APPROVED** [review]

**Comment by sebastiangrimberg** [comment]
> I added two "TODO" items in the initial PR post.

Done in 574980a. Thank you!

File: fem/restriction.hpp [hunk 3]
```diff
@@ -21,20 +21,10 @@ namespace mfem
...
 
-/// Abstract base class that defines an interface for element restrictions.
-class ElementRestrictionOperator : public Operator
-{
-public:
-   /// @brief Add the E-vector degrees of freedom @a x to the L-vector degrees
-   /// of freedom @a y.
-   void AddMultTranspose(const Vector &x, Vector &y,
-                         const double a = 1.0) const override = 0;
-};
```
Comment (YohannDudouit) [inline fem/restriction.hpp:32]: Was there no need for this class after all, removing this class seems a bit unrelated to the purpose of this PR.

**Review by YohannDudouit — APPROVED** [review]
Overall I think this is a big organization improvement, I would personally go even further, I don't like files with more than a few hundreds of lines. For instance, I would personally  replace `_` with deeper hierarchy, for instance `bilininteg_diffusion_pa.cpp` would be in `integ/bilininteg/diffusion/pa.hpp`.
I'm a bit confused by the removal of the `ElementRestrictionOperator`, was it added unnecessarily?

File: fem/restriction.hpp (hunk 3 above)
Comment (pazner) [reply fem/restriction.hpp:32]: The `ElementRestrictionOperator` was needed for `AddMultTranspose` before PR #3075, which has made this abstract base class obsolete (the interface is now in `Operator` itself).

I don't see the harm in leaving it in this PR

Comment (YohannDudouit) [reply fem/restriction.hpp:32]: I am not against having this in this PR (though having atomic PR is always better), but can you remind me why it was introduced, and why you're judging now that it's unneeded?

Comment (v-dobrev) [reply fem/restriction.hpp:32]: Logically, it is good to have a common base class for all element-restriction classes, even if that does not do anything specific at the moment (in the future we may want to add methods, e.g. to query the E-vector layout when using mixed meshes or variable orders). Also, removing the base class changes the API in a way that is not backward compatible -- this needs to be documented in the CHANGELOG if you decide to keep this change.

Comment (pazner) [reply fem/restriction.hpp:32]: How about we decouple this change from the rest of the PR since it's orthogonal. We can have a separate PR or discussion about this proposed change.

Comment (sebastiangrimberg) [reply fem/restriction.hpp]: Agreed, done in d2e2b5d. See #3599 for the separate PR.

**Comment by sebastiangrimberg** [comment]
@mlstowell Sorry to bother you but is there any chance you can take a look at this when you get a few minutes? Thanks a lot!

**Review by mlstowell — APPROVED** [review]

**Comment by pazner** [comment]
Merged in `next` for testing...

File: fem/integ/bilininteg_diffusion_kernels.hpp [hunk 4]
```diff
@@ -9,28 +9,29 @@
...
 // OCCA 2D Assemble kernel
 #ifdef MFEM_USE_OCCA
-static void OccaPADiffusionSetup2D(const int D1D,
-                                   const int Q1D,
-                                   const int NE,
-                                   const Array<double> &W,
-                                   const Vector &J,
-                                   const Vector &C,
-                                   Vector &op)
+MFEM_HOST_DEVICE inline
```
[the remark this answers is not in the document]
Comment (sebastiangrimberg) [reply fem/integ/bilininteg_diffusion_kernels.hpp:27]: I think `inline` is required otherwise we're going to get multiple definition errors (function definition outside of a class body in a header file is going to be defined once per translation unit). As for `MFEM_HOST_DEVICE`, you are right I think since the kernels are launched from inside these functions.

[the remark this answers is not in the document]
Comment (sebastiangrimberg) [reply fem/integ/bilininteg_diffusion_kernels.hpp:27]: As it stands, the reorganization created the `*_kernels.hpp` files to include all the free functions required for integrators (many of which are shared between integrators).

These new files could be separated into `*_kernels.hpp` and `*_kernels.cpp` to save a few `inline`s, but the impact will be minimal since most are template functions and I don't think we are trying to introduce any explicit instantiation. So, I don't really see much value in splitting and keeping them header only seems like a good approach to me.

If alternatively you are suggesting moving some of the functions from the `*_kernels.hpp` files back into the `bilininteg_*.cpp` files, I think this starts to break the separation between integrator class member functions (`AssemblePA`, `AddMultPA`, ...) and the free (often shared) helper functions which launch kernels.

[the remark this answers is not in the document]
Comment (sebastiangrimberg) [reply fem/integ/bilininteg_diffusion_kernels.hpp]: As a start, see [9c0eae5](https://github.com/mfem/mfem/pull/3526/commits/9c0eae5e3e43f3f80d34ac8cbdfa9ab56b6215f2) for the `MFEM_HOST_DEVICE` fix.

[the remark this answers is not in the document]
Comment (pazner) [reply fem/integ/bilininteg_diffusion_kernels.hpp]: OK, sounds fine to me.

Are you able to build with CUDA? I'm getting warnings like
```
cc1plus: warning: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' writing between 8 and 17179869176 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
```

[the remark this answers is not in the document]
Comment (sebastiangrimberg) [reply fem/integ/bilininteg_diffusion_kernels.hpp]: I don't have a test environment set up currently but could do that if needed. Is there any more context to the warnings to try to debug?

File: fem/integ/bilininteg_hcurl_kernels.hpp [hunk 5]
```diff
@@ -0,0 +1,3869 @@
...
+   auto Bo = Reshape(bo.Read(), Q1D, D1D-1);
+   auto Bc = Reshape(bc.Read(), Q1D, D1D);
+   auto Gc = Reshape(gc.Read(), Q1D, D1D);
+   auto op = Reshape(pa_data.Read(), Q1D, Q1D, Q1D, symmetric ? 6 : 9, NE);
+   auto X = Reshape(x.Read(), 3*(D1D-1)*D1D*D1D, NE);
+   auto Y = Reshape(y.ReadWrite(), 3*(D1D-1)*D1D*D1D, NE);
+
+   const int s = symmetric ? 6 : 9;
+
+   auto device_kernel = [=] MFEM_HOST_DEVICE (int e)
```
Comment (pazner) [inline fem/integ/bilininteg_hcurl_kernels.hpp:2019]:
```suggestion
   auto device_kernel = [=] MFEM_DEVICE (int e)
```

This (and the same change made in other places) is the cause of the warnings referenced above.

@sebastiangrimberg, do you know of a good way to do a full diff of the moved function bodies to make sure that inadvertent edits like this don't get merged?

Comment (sebastiangrimberg) [reply fem/integ/bilininteg_hcurl_kernels.hpp]: Looks like I was too aggressive with my find-and-replace in [9c0eae5](https://github.com/mfem/mfem/pull/3526/commits/9c0eae5e3e43f3f80d34ac8cbdfa9ab56b6215f2). Fixed this in [1a614d5](https://github.com/mfem/mfem/pull/3526/commits/1a614d51a2d98b9c0d13458e2339fcac9f41ffa3).

It would be nice to compare function-by-function changes in this branch, but unfortunately no I don't know any way to do it easily. I just parse the full diff to check my work and up to this point have done this pretty stringently to make sure things are correct. The mistake from [1a614d5](https://github.com/mfem/mfem/pull/3526/commits/1a614d51a2d98b9c0d13458e2339fcac9f41ffa3) is really touching anything inside of functions but I understand the desire to double check everything. I will give things another review now just to make extra sure, but I guess at the end of the day we will have to trust strong unit testing + CI testing.

Comment (pazner) [reply fem/integ/bilininteg_hcurl_kernels.hpp]: Thanks for the fix. It builds cleanly with CUDA for me now… I'll re-merge in next.

Please do check over the changes again and see if anything else slipped through the cracks.

**Comment by pazner** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
Previously, methods like `PADiffusionSetup2D<2>` were defined once and used multiple times. The modifications proposed here make the same method `inline` and so its use in different translation units will result in multiple copies. For small low level functions this makes sense, however, for methods like `PADiffusionSetup2D<2>` this just slows down compilation and inflates the size of the generated code unnecessarily.

There are many methods that are modified by the PR in the same way. I think this issue needs to be addressed before merging.

**Comment by v-dobrev** [comment]
I also see changes that should not be hidden in a PR like this. For example, currently in `master` we have this https://github.com/mfem/mfem/blob/d0e75c8d0d8b6def5ae22a38e08e81dd8e38c0d8/fem/bilininteg_hcurl.cpp#L311-L318
which in this PR has been edited to become https://github.com/mfem/mfem/blob/3adc51440f9cb6e548cb0e1ce1096f8a10700fdd/fem/integ/bilininteg_hcurl_kernels.hpp#L179-L183

There is a reason the code in `master` is written that way -- to workaround an issue with `nvcc`, see #852. Making changes like this while reorganizing the code is a bad idea. I only found this because I spent quite some time tracking the differences and I suspect there are other edits in the code that deviate from simple copy-paste changes.

I'll continue looking at the differences, however, we need to figure out a better way to make this type of reorg changes -- specifically to make it easy for reviewers to see small edits that look innocent but can cause problems.

**Comment by pazner** [comment]
Thanks for looking into this closely @v-dobrev — good catch. It is possible that the change you point out is actually fine (since `T_Q1D` is a template parameter — and I believe `static constexpr` outside the lambda also works to avoid the issue in #852), but I completely agree that these kinds of changes should be evaluated **outside** of the context of a code reorganization.

**Comment by sebastiangrimberg** [comment]
Hi @v-dobrev, it was certainly not my intention to hide any changes in this PR so hopefully I can give an explanation of why I thought this change was actually normal as part of the reorganization. Along the lines of what @pazner points out, I think this is actually a different circumstance than the PR #852 you linked, which is dealing with the diffusion kernels and their allowing for runtime specification of the number of element nodes and quadrature points via the function parameters `d1d` and `q1d` when the template parameters `T_D1D` and `T_Q1D` are defaults. In that case, there was some trickery done in order to get the compiler to recognize when in fact the template parameters were nonzero at compile time and perform the subsequent optimizations.

In the case you linked, however, that's not the situation and the change I made which adds maximum size default template parameters to `SmemPAHcurlMassAssembleDiagonal3D` and `SmemPAHcurlMassApply3D` rather than assigning them inside the function, was to match the style of all of the rest of the templated kernels in `bilininteg_hcurl_kernels.hpp`, including `PACurlCurlAssembleDiagonal3D`, `SmemPACurlCurlAssembleDiagonal3D`, `PACurlCurlApply3D`, `SmemPACurlCurlApply3D`, and `PAHcurlL2Apply3D` and `SmemPAHcurlL2Apply3D` (and their transposes). I apologize if this appears to be adding on extraneous changes to a reorganization PR, but this I see this as an organizational change for code consistency (with no difference in compiled code or performance unlike #852) as these were just two different ways of doing the same thing and now that we no longer need the list of explicit instantiations used for these two methods after the reorganization, this was a purely syntax change to match the existing other kernels in the same file. I was hoping this would avoid any confusion in the future.

I apologize again, and if you'd like I can remove the default template parameters on these two functions to go back to the previous code. I do not believe there are any other changes I made which could be qualified as "hidden changes" in this PR but am always willing to reconsider and listen to editors' requests. I guess [this commit](https://github.com/mfem/mfem/pull/3526/commits/cb962cf1db66709d9f3f323013ab853c6c418210) might be deemed "hidden" but I hoped the approval by reviewers had suggested it was OK as part of this reorganization. Thanks again for your help and looking forward to getting these improvements ready to merge to `master`.

**Comment by sebastiangrimberg** [comment]
> Previously, methods like `PADiffusionSetup2D<2>` were defined once and used multiple times. The modifications proposed here make the same method `inline` and so its use in different translation units will result in multiple copies. For small low level functions this makes sense, however, for methods like `PADiffusionSetup2D<2>` this just slows down compilation and inflates the size of the generated code unnecessarily.

This is in line with @pazner's comment [here](https://github.com/mfem/mfem/pull/3526#discussion_r1213546801). Indeed it should speed up compilation time to move these into `bilininteg_*_kernels.cpp` files where possible. This is done in commit 51e24f9, hopefully this resolves the issue but please let me know if this is not what you had in mind.

File: fem/integ/bilininteg_hcurlhdiv_kernels.hpp [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+   const int D1Dtest = T_D1D_TEST ? T_D1D_TEST : d1dtest;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+   auto Bo = Reshape(bo.Read(), Q1D, D1D-1);
+   auto Bc = Reshape(bc.Read(), Q1D, D1D);
+   auto Bot = Reshape(bot.Read(), D1Dtest-1, Q1D);
+   auto Bct = Reshape(bct.Read(), D1Dtest, Q1D);
+   auto Gct = Reshape(gct.Read(), D1D, Q1D);
+   auto op = Reshape(pa_data.Read(), Q1D, Q1D, Q1D, 6, NE);
+   auto X = Reshape(x.Read(), 3*(D1Dtest-1)*(D1Dtest-1)*D1D, NE);
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurlhdiv_kernels.hpp:476]: This is the same as in `master`: https://github.com/mfem/mfem/blob/19c16996d5117a13cc7242d7bfbe88342053285f/fem/bilininteg_hcurl.cpp#L4608

but it does look like a bug/typo: `D1D` -> `D1Dtest`.

File: fem/integ/bilininteg_hcurlhdiv_kernels.hpp [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+   auto Bo = Reshape(bo.Read(), Q1D, D1D-1);
+   auto Bc = Reshape(bc.Read(), Q1D, D1D);
+   auto Bot = Reshape(bot.Read(), D1Dtest-1, Q1D);
+   auto Bct = Reshape(bct.Read(), D1Dtest, Q1D);
+   auto Gc = Reshape(gc.Read(), Q1D, D1D);
+   auto op = Reshape(pa_data.Read(), Q1D, Q1D, Q1D, 6, NE);
+   auto X = Reshape(x.Read(), 3*(D1D-1)*D1D*D1D, NE);
+   auto Y = Reshape(y.ReadWrite(), 3*(D1Dtest-1)*(D1Dtest-1)*D1D, NE);
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurlhdiv_kernels.hpp:110]: Same here: `D1D` -> `D1Dtest`. Same code in `master`: https://github.com/mfem/mfem/blob/19c16996d5117a13cc7242d7bfbe88342053285f/fem/bilininteg_hcurl.cpp#L4253

Comment (sebastiangrimberg) [reply fem/integ/bilininteg_hcurlhdiv_kernels.hpp:110]: Yup, great catches. I had found and fixed similar bugs/typos in d9965bf69a463d997d8639ea80ae6907db35c07c but missed these. I am away for a long weekend but will fix these as well ASAP, or you can feel free to make the change.

Comment (v-dobrev) [reply fem/integ/bilininteg_hcurlhdiv_kernels.hpp:110]: I noticed these because of d9965bf6. 😄

File: fem/integ/bilininteg_hcurlhdiv_kernels.hpp (hunk 6 above)
Comment (v-dobrev) [reply fem/integ/bilininteg_hcurlhdiv_kernels.hpp]: Fixed in 8c93cb49.

**Comment by v-dobrev** [comment]
Another set of changes that I see is that a number of comment were cut short, e.g. here: https://github.com/mfem/mfem/blob/8c93cb4947e0bf6e523ea2f110520816b7cceb02/fem/integ/bilininteg_hcurlhdiv_kernels.hpp#L28-L29 compared to the original: https://github.com/mfem/mfem/blob/19c16996d5117a13cc7242d7bfbe88342053285f/fem/bilininteg_vectorfe.cpp#L308-L311 (the renaming of the function is fine.)

I'm not sure if it makes a big difference but in most cases more comments is better than less. 😄

**Comment by v-dobrev** [comment]
@pazner, I don't have any further comments about this PR -- it just needs to be re-merged in `next` to test the latest changes.

**Comment by pazner** [comment]
Re-merged in `next` for testing...

**Comment by sebastiangrimberg** [comment]
> Another set of changes that I see is that a number of comments were cut short, e.g. here:
> 
> https://github.com/mfem/mfem/blob/8c93cb4947e0bf6e523ea2f110520816b7cceb02/fem/integ/bilininteg_hcurlhdiv_kernels.hpp#L28-L29
> 
> compared to the original:
> https://github.com/mfem/mfem/blob/19c16996d5117a13cc7242d7bfbe88342053285f/fem/bilininteg_vectorfe.cpp#L308-L311
> 
> (the renaming of the function is fine.)
> I'm not sure if it makes a big difference but in most cases more comments is better than less. 😄

Certainly agree that more comments are useful in understanding these partial assembly methods as sometimes the code can get quite lengthy and isn't so straightforward to understand 😄 . In this case and I believe most files here I tried to keep the comments in header files short and consistent while the extended comments on the actual implementation were moved to the function definitions (here, in `bilininteg_hcurlhdiv_kernels.cpp`). Anyways, thanks again for your reviews and help on this @v-dobrev and @pazner!

**Comment by pazner** [comment]
Thank you for the contribution @sebastiangrimberg! Thank you for reviewing @v-dobrev, @YohannDudouit, @mlstowell!
